### PR TITLE
Split multiple distinct aggregations to sub queries

### PR DESCRIPTION
**Pull request by Dith3r**
## Description

Introduce new rule that splits distinct aggregations on different arguments to sub-queries and joins the grouped
results using grouping keys if any.
This allows `SingleDistinctAggregationToGroupBy` to kick in and improve parallelism and performance significantly when the grouped query is cheap to duplicate.

Benchmarks
obraz
Some queries (simple aggregation on top of table scan, with low cardinality group by) can be significantly improved like 5s vs 40s using MarkDistinct

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
# General
* Improve performance of queries with multiple distinct aggregations. ({issue}`22355`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DistinctAggregationStrategyChooser.java [hunk 1]
```diff
@@ -17,9 +17,17 @@
 import io.trino.cost.PlanNodeStatsEstimate;
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DistinctAggregationStrategyChooser.java:19]: Change the commit message DistinctAggregationController  -> DistinctAggregationStrategyChooser

**Review by lukasz-stec — APPROVED** [review]
lgtm

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationsToSubqueries.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    private JoinNode buildJoin(PlanNode left, List<Symbol> leftJoinSymbols, PlanNode right, List<Symbol> rightJoinSymbols, Context context)
+    {
+        checkArgument(leftJoinSymbols.size() == rightJoinSymbols.size());
+        List<EquiJoinClause> criteria = IntStream.range(0, leftJoinSymbols.size())
+                .mapToObj(i -> new EquiJoinClause(leftJoinSymbols.get(i), rightJoinSymbols.get(i)))
+                .collect(toImmutableList());
+
+        // TODO: we dont need dynamic filters for this join at all. We could add skipDf field to the JoinNode and make use of it in PredicatePushDown
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationsToSubqueries.java:190]: I didn't get what this TODO means, why would we "add skipDf field to the JoinNode and make use of it in PredicatePushDown" ? If there is no DF to be pushed down, then PredicatePushDown will do nothing.

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestMultipleDistinctAggregationsToSubqueries.java [hunk 3]
```diff
@@ -0,0 +1,1694 @@
...
+                1);
+        planTester.createCatalog(
+                session.getCatalog().orElseThrow(),
+                new TpchConnectorFactory(1),
+                ImmutableMap.of());
+        return new RuleTester(planTester);
+    }
+
+    private static class DelegatingMetadata
+            implements Metadata
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestMultipleDistinctAggregationsToSubqueries.java:1180]: Can we just extend `AbstractMockMetadata` and avoid implementing entire interface ?

File: core/trino-main/src/main/java/io/trino/testing/PlanTester.java [hunk 4]
```diff
@@ -302,15 +303,20 @@ public class PlanTester
...
+        return new PlanTester(defaultSession, identity(), nodeCountForStats);
     }
 
-    private PlanTester(Session defaultSession, int nodeCountForStats)
+    public static PlanTester create(Session defaultSession, Function<Metadata, Metadata> metadataDecorator, int nodeCountForStats)
+    {
+        return new PlanTester(defaultSession, metadataDecorator, nodeCountForStats);
+    }
+
+    private PlanTester(Session defaultSession, Function<Metadata, Metadata> metadataDecorator, int nodeCountForStats)
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/testing/PlanTester.java:319]: instead of `Function<Metadata, Metadata> metadataDecorator`, you can take `Optional<Metadata> metadata` as parameter

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationsToSubqueries.java (hunk 2 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationsToSubqueries.java:190]: The join created here joins multiple sources that are guaranteed to have the same values in the columns that are part of join condition. This makes DF not beneficial at all so it just adds overhead. We know this here but this knowledge is lost in the `PredicatePushDown` where the DF is created

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationsToSubqueries.java:190]: okay, i think see the point now

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 5]
```diff
@@ -667,6 +668,12 @@ public TableScanBuilder setUpdateTarget(boolean updateTarget)
             return this;
         }
 
+        public TableScanBuilder setNodeId(PlanNodeId id)
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/test/PlanBuilder.java:671]: why was this needed ? usually plan matcher tests are not dependant on plan node id

Comment (Dith3r) [reply core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/test/PlanBuilder.java:671]: it is needed for stats override

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 6]
```diff
@@ -1814,6 +1814,14 @@ default OptionalInt getMaxWriterTasks(ConnectorSession session)
         return OptionalInt.empty();
     }
 
+    /**
+     * @return whether connector supports reading a single column from the give tableHandle, without reading the entire table.
+     */
+    default boolean isColumnarTableScan(ConnectorSession session, ConnectorTableHandle tableHandle)
```
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: `ConnectorTableHandle` can represent a table or a relation (see JDBC pushdowns).  Also, selected columns are conveyed to the connector via applyProjection. Is it possible to rephrase the javadoc with these in mind?

Also, given applyProjection was already invoked, why does the engine care about other columns that are not going to be read? Maybe knowing this would help formulate the javadoc best.

[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: This method should return true only for columnar data formats that allow efficient reads of single columns i.e. reading column `a` separately from column `b` requires more or less the same effort as reading columns `a` and `b` together.
For this to be useful, projection pushdown also needs to be supported by the connector.
@findepi Maybe you can suggest a better javadoc with this in mind?
@Dith3r there is a typo there _give_ -> _given_

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: SPI does not (and should not) know about file formats.
What is the SPI way to read `a` column separate from `b` , or read them together?
My guess is that it's very implicit and related to lazy blocks being returned from the page source. Is this correct?

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: > My guess is that it's very implicit and related to lazy blocks being returned from the page source. Is this correct?

There are multiple levels of "columnar" reads:
1. Physically reading columns separately (that usually not happens as reads are merged)
2. Decoding columns separately.

This SPI is telling that cost of reading `a, b` is more or less sum of cost of reading `a` and `b` separately (when pruning projections are applied), so maybe that's better Javadoc for it

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: We don't have a SPI method that takes a table handle (supposed to return `a,b`) and returns two separate table handles, one returning `a` and the other `b` column.

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: > We don't have a SPI method that takes a table handle (supposed to return a,b) and returns two separate table handles, one returning a and the other b column.

This is what this PR is effectively doing

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: what about

```
@return true if reading subset of columns from given table separately from reading a compliment of the subset has similar or better
performance as reading this table.
```

?

[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: Looks good to me with a small correction. You probably meant _complement_?

[the remark this answers is not in the document]
Comment (martint) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: This is a poor proxy for indicating whether a read is cheaper. The fact that we have to explain it in the javadocs means the method semantics are not very well specified -- and the explanation has nothing to do with the name of the method.

If the goal is to establish the cost of reading a set of columns, we should model it using cost estimations like we do elsewhere. I haven't thought enough about it, but that's the direction I would take. Maybe do a comparison between a given table handle that encompasses columns `a` and `b` (from the examples above) and the sum of the costs of pushing `a` and `b` separately into the table scan.

[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: The goal is not to establish a generic cost of reading but rather to determine if splitting a read into two reads does not add much overhead. That is what basically only columnar storage provides, hence the name.
Having the cost would of course work here but it is too complex for the use case and would make the implementation harder than necessary. For example, we don't need to compare the cost of reading two different columns here.

[the remark this answers is not in the document]
Comment (martint) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: > but rather to determine if splitting a read into two reads does not add much overhead. 

That's too specific for the SPI. And still, whether the read is columnar is not necessarily indicative of that. There are many other factors involved, including whether processing metadata is expensive, how many rows are read, how many columns there are in the table vs how many are being read, I/O costs if the columns are scattered, how much data is being read from underlying storage vs decompressed vs decoded, etc.

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: Maybe you could add `Estimate dataSize` along with `io.trino.spi.statistics.TableStatistics#rowCount`. Could be unknown all connectors than "columnar" connectors. For "columnar" connectors it could be sum of dataSize of all involved columns. 
This way you could compare data size involved in table scan of `AB` with scan of `A` and `B` separately. 
wdyt?

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: We may not know the data sizes until eg split generation, or even later.
I am OK with "determine if splitting a read into two reads does not add much overhead" concept.
I see how being more precise could be helpful, but it could force implementors into faking answers.

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: > We may not know the data sizes until eg split generation, or even later.

Sure, it would be an estimate same as for column data size

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: In Iceberg, estimate for a column size requires walking the manifest files and then processing data file information from them. Depending on number of files in a table, this is not cheap and definitely not free. Meanwhile, Iceberg can blindly answer the "whether splitting a read into two reads does not add much overhead" question.

[the remark this answers is not in the document]
Comment (Dith3r) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: Do we have any conclusion which path is right one?

[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: One aspect we didn't discuss here is that with JDBC connectors there may be an expensive to compute join/aggregation pushed into the table handle and we don't want to repeat that computation only to project different column out of it for each sub-query.
That scenario is more problematic compared to wasted read from csv/text/avro etc.
How about we name this a bit differently
```
public boolean allowSplittingReadIntoMultipleSubQueries(ConnectorSession session, ConnectorTableHandle tableHandle)
``` 
A bit clunky but makes it clearer that it's more about making a judgement whether we want to split a read from table handle rather than about columnar/row orientation of data.

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1820]: makes sense to me. wdyt @martint ?

File: core/trino-main/src/main/java/io/trino/metadata/Metadata.java [hunk 7]
```diff
@@ -840,6 +840,15 @@ default boolean isMaterializedView(Session session, QualifiedObjectName viewName
      */
     OptionalInt getMaxWriterTasks(Session session, String catalogName);
 
+    /**
+     * Workaround to lack of statistics about IO and CPU operations performed by the connector.
+     * In the long term, this should be replaced by improvements in the cost model.
+     *
+     * @return true if reading a subset of columns from a given table separately from reading a complement of the subset has similar or better
+     * performance as reading table.
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/metadata/Metadata.java:848]:
```
@return true if the cumulative cost of splitting a read of the specified tableHandle into multiple reads, each of which projects a subset of the required columns, is not significantly more than the cost of reading the specified tableHandle
```

**Review by raunaqmorarka — APPROVED** [review]

File: testing/trino-tests/src/test/java/io/trino/tests/TestDistinctToSubqueriesAggregations.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import io.trino.tpch.TpchTable;
+
+public class TestDistinctToSubqueriesAggregations
+        extends AbstractTestAggregations
+{
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        // using memory connector, because it supports efficient single column reads
```
Comment (raunaqmorarka) [inline testing/trino-tests/src/test/java/io/trino/tests/TestDistinctToSubqueriesAggregations.java:29]: `efficient single column reads` -> `enables ConnectorMetadata#allowSplittingReadIntoMultipleSubQueries`

File: docs/src/main/sphinx/admin/properties-optimizer.md [hunk 9]
```diff
@@ -54,6 +54,8 @@ This strategy will perform poorly if the number of distinct grouping keys is sma
 or for mix of distinct and non-distinct aggregations.
 `PRE_AGGREGATE` Computes distinct aggregations using a combination of aggregation
 and pre-aggregation steps.
+`SPLIT_TO_SUBQUERIES` Splits the aggregation input to independent sub-queries,
+where each subquery computes single distinct aggregation thus improving parallelism
```
Comment (raunaqmorarka) [inline docs/src/main/sphinx/admin/properties-optimizer.md:58]: Please make changes to display the strategies as a list. Right now they are being rendered as a single block of text.

File: core/trino-main/src/test/java/io/trino/connector/MockConnectorFactory.java [hunk 10]
```diff
@@ -149,6 +149,7 @@ public class MockConnectorFactory
 
     private final WriterScalingOptions writerScalingOptions;
     private final Supplier<Set<ConnectorCapabilities>> capabilities;
+    private final boolean supportsSingleColumnReads;
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/connector/MockConnectorFactory.java:152]: rename to allowSplittingReadIntoMultipleSubQueries

File: core/trino-main/src/test/java/io/trino/connector/MockConnectorFactory.java [hunk 11]
```diff
@@ -474,6 +478,7 @@ public static final class Builder
         private BiFunction<ConnectorSession, ConnectorTableExecuteHandle, Optional<ConnectorTableLayout>> getLayoutForTableExecute = (session, handle) -> Optional.empty();
         private WriterScalingOptions writerScalingOptions = WriterScalingOptions.DISABLED;
         private Supplier<Set<ConnectorCapabilities>> capabilities = ImmutableSet::of;
+        private boolean supportsSingleColumnReads;
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/connector/MockConnectorFactory.java:481]: rename to allowSplittingReadIntoMultipleSubQueries

File: testing/trino-tests/src/test/java/io/trino/tests/TestDistinctToSubqueriesAggregations.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import io.trino.tpch.TpchTable;
+
+public class TestDistinctToSubqueriesAggregations
+        extends AbstractTestAggregations
+{
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        // using memory connector, because it supports enables ConnectorMetadata#allowSplittingReadIntoMultipleSubQueries
```
Comment (raunaqmorarka) [inline testing/trino-tests/src/test/java/io/trino/tests/TestDistinctToSubqueriesAggregations.java:29]: skip `supports`

File: docs/src/main/sphinx/admin/properties-optimizer.md [hunk 13]
```diff
@@ -43,18 +43,27 @@ create them.
...
-- **Allowed values:** `AUTOMATIC`, `MARK_DISTINCT`, `SINGLE_STEP`, `PRE_AGGREGATE`
+- **Allowed values:** `AUTOMATIC`, `MARK_DISTINCT`, `SINGLE_STEP`, `PRE_AGGREGATE`, `SPLIT_TO_SUBQUERIES`
 - **Default value:** `AUTOMATIC`
 - **Session property:** `distinct_aggregations_strategy`
 
 The strategy to use for multiple distinct aggregations.
-`SINGLE_STEP` Computes distinct aggregations in single-step without any pre-aggregations.
+- `SINGLE_STEP` Computes distinct aggregations in single-step without any pre-aggregations.
 This strategy will perform poorly if the number of distinct grouping keys is small.
+<<<<<<< Updated upstream
```
Comment (raunaqmorarka) [inline docs/src/main/sphinx/admin/properties-optimizer.md:53]: please fix merge conflicts

**Review by raunaqmorarka — APPROVED** [review]
